File: webhooks/csharp/Utils/QuerySqliteDb.cs
using System.Runtime.CompilerServices;
using Microsoft.Data.Sqlite;

namespace WebhooksCodelab.Utils;

/// <summary>Prints emails stored by the DB for the webhook codelab.</summary>
public static class QuerySqliteDb
{
    /// <summary>The local database file, created by the create-db step.</summary>
    public static string DatabasePath { get; } = LocateDatabase();

    // Exercise 3: Use the recipient argument in get_recipient_email_history to
    // retrieve only the selected recipient's records.
    /// <summary>Return stored emails in insertion order, optionally for one recipient.</summary>
    public static async Task<List<Dictionary<string, object?>>> FetchEmailsAsync(string? recipient = null)
    {
        await using var connection = await OpenReadOnlyDatabaseAsync();
        await using var command = connection.CreateCommand();

        command.CommandText = """
            SELECT *
            FROM emails
            """;
        if (recipient is not null)
        {
            command.CommandText += "\nWHERE recipients = $recipient";
            command.Parameters.AddWithValue("$recipient", recipient);
        }
        command.CommandText += "\nORDER BY id";

        return await ReadRowsAsync(command);
    }

    // Exercise 3: Use these rows in summarize_email_activity to total the emails and return
    // each recipient's activity. Each row's email_count is a long.
    /// <summary>Return one email-count row per recipient, ordered by activity.</summary>
    public static async Task<List<Dictionary<string, object?>>> FetchEmailActivityAsync()
    {
        await using var connection = await OpenReadOnlyDatabaseAsync();
        await using var command = connection.CreateCommand();

        command.CommandText = """
            SELECT
                recipients AS recipient,
                COUNT(*) AS email_count
            FROM emails
            GROUP BY recipients
            ORDER BY email_count DESC, recipient
            """;

        return await ReadRowsAsync(command);
    }

    /// <summary>Print email rows in an aligned table with database field headings.</summary>
    public static void PrintEmails(List<Dictionary<string, object?>> emails)
    {
        var fieldNames = emails[0].Keys.ToArray();
        var rows = emails
            .Select(email => fieldNames.Select(field => email[field]?.ToString() ?? "").ToArray())
            .ToList();
        var columnWidths = fieldNames
            .Select((field, index) => Math.Max(field.Length, rows.Max(row => row[index].Length)))
            .ToArray();

        string FormatRow(string[] values) =>
            string.Join(" | ", values.Select((value, index) => value.PadRight(columnWidths[index])));

        Console.WriteLine(FormatRow(fieldNames));
        Console.WriteLine(string.Join("-+-", columnWidths.Select(width => new string('-', width))));
        foreach (var row in rows)
        {
            Console.WriteLine(FormatRow(row));
        }
    }

    /// <summary>Query the codelab database and print readable email records.</summary>
    public static async Task RunAsync()
    {
        List<Dictionary<string, object?>> emails;
        try
        {
            emails = await FetchEmailsAsync();
        }
        catch (Exception error) when (error is FileNotFoundException or SqliteException)
        {
            Console.Error.WriteLine($"Unable to query stored emails: {error.Message}");
            Environment.ExitCode = 1;
            return;
        }

        if (emails.Count == 0)
        {
            Console.WriteLine($"No emails found in {DatabasePath}");
        }
        else
        {
            PrintEmails(emails);
        }
    }

    /// <summary>
    /// Helper function that opens a read-only connection to the codelab database.
    ///
    /// Checking the path first gives a clear error when the database has not
    /// been created yet, instead of a generic SQLite "unable to open" error.
    /// </summary>
    private static async Task<SqliteConnection> OpenReadOnlyDatabaseAsync()
    {
        if (!File.Exists(DatabasePath))
        {
            throw new FileNotFoundException(
                $"Database not found at {DatabasePath}. Run the create-db step first.");
        }

        // Read-only mode prevents this query helper and the agent tools that use
        // it from modifying the codelab database.
        var connectionString = new SqliteConnectionStringBuilder
        {
            DataSource = DatabasePath,
            Mode = SqliteOpenMode.ReadOnly,
            // A pooled connection keeps its file open, so a long-running serve-agent step
            // would keep reading a database file even after it was deleted and re-created.
            Pooling = false,
        }.ToString();
        var connection = new SqliteConnection(connectionString);
        await connection.OpenAsync();
        return connection;
    }

    /// <summary>Read every result row as a dictionary keyed by column name.</summary>
    private static async Task<List<Dictionary<string, object?>>> ReadRowsAsync(SqliteCommand command)
    {
        await using var reader = await command.ExecuteReaderAsync();

        var rows = new List<Dictionary<string, object?>>();
        while (await reader.ReadAsync())
        {
            var row = new Dictionary<string, object?>();
            for (var index = 0; index < reader.FieldCount; index++)
            {
                row[reader.GetName(index)] = reader.IsDBNull(index) ? null : reader.GetValue(index);
            }
            rows.Add(row);
        }
        return rows;
    }

    // Resolved from this source file's folder, as the Python version does, rather
    // than the build output folder, so every step uses the same database file.
    private static string LocateDatabase([CallerFilePath] string sourceFile = "") =>
        Path.Combine(Path.GetDirectoryName(sourceFile)!, "webhook_codelab_storage.db");
}

File: webhooks/typescript/utils/querySqliteDb.ts
/** Prints emails stored by the DB for the webhook codelab. */

import { statSync } from "node:fs";
import path from "node:path";
import { DatabaseSync, type SQLOutputValue } from "node:sqlite";

/** The local database file, created by the create-db step next to this file. */
export const DATABASE_PATH = path.join(import.meta.dirname, "webhook_codelab_storage.db");

/** One database row, keyed by column name. */
export type Row = Record<string, SQLOutputValue>;

/**
 * Open a read-only connection to the codelab database. Checking the path first gives a clear error
 * when the database has not been created yet.
 */
function openReadOnlyDatabase(): DatabaseSync {
  if (!statSync(DATABASE_PATH, { throwIfNoEntry: false })?.isFile()) {
    throw new Error(
      `Database not found at ${DATABASE_PATH}. Run the create-db step (npm run create-db) first.`,
    );
  }

  // Read-only mode prevents this query helper and the agent tools that use it from modifying the
  // codelab database.
  return new DatabaseSync(DATABASE_PATH, { readOnly: true });
}

// Exercise 3: Use the recipient argument in get_recipient_email_history to
// retrieve only the selected recipient's records.
/** Return stored emails in insertion order, optionally for one recipient. */
export function fetchEmails(recipient?: string): Row[] {
  let query = `
    SELECT *
    FROM emails
    `;
  const parameters: string[] = [];
  if (recipient !== undefined) {
    query += "WHERE recipients = ?\n";
    parameters.push(recipient);
  }
  query += "ORDER BY id";

  const database = openReadOnlyDatabase();
  try {
    return database.prepare(query).all(...parameters);
  } finally {
    database.close();
  }
}

// Exercise 3: Use these rows in summarize_email_activity to total the emails and return each
// recipient's activity. Each row's email_count is a number, although its type is SQLOutputValue.
/** Return one email-count row per recipient, ordered by activity. */
export function fetchEmailActivity(): Row[] {
  const query = `
    SELECT
      recipients AS recipient,
      COUNT(*) AS email_count
    FROM emails
    GROUP BY recipients
    ORDER BY email_count DESC, recipient
    `;

  const database = openReadOnlyDatabase();
  try {
    return database.prepare(query).all();
  } finally {
    database.close();
  }
}

/** Print email rows in an aligned table with database field headings. */
export function printEmails(emails: Row[]): void {
  const fieldNames = Object.keys(emails[0] ?? {});
  const rows = emails.map((email) => fieldNames.map((field) => String(email[field])));
  // Each column is as wide as its longest value, counting the heading as a value.
  const columnWidths = fieldNames.map((field) =>
    Math.max(field.length, ...emails.map((email) => String(email[field]).length)),
  );

  const formatRow = (values: string[]): string =>
    values.map((value, index) => value.padEnd(columnWidths[index] ?? 0)).join(" | ");

  console.log(formatRow(fieldNames));
  console.log(columnWidths.map((width) => "-".repeat(width)).join("-+-"));
  for (const row of rows) {
    console.log(formatRow(row));
  }
}

/** Query the codelab database and print readable email records. */
function main(): number {
  let emails: Row[];
  try {
    emails = fetchEmails();
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Unable to query stored emails: ${message}`);
    return 1;
  }

  if (emails.length === 0) {
    console.log(`No emails found in ${DATABASE_PATH}`);
  } else {
    printEmails(emails);
  }
  return 0;
}

// Run only as the query-db step, not when the agent's tools import this file.
if (import.meta.main) {
  process.exitCode = main();
}
